Accessibility API for CustomPainter

yjbanov opened the pull request:
Summary:

- Add `key` field to `SemanticsNode`, while moving key into `foundation` library so it can be used by the render layer.
- Introduce `SemanticsProperties` and move many of the `Semantics` fields into it.
- Introduce `CustomPaintSemantics` - a `SemanticsNode` prototype created by `CustomPainter`.
- Introduce `semanticsBuilder` and `shouldRebuildSemantics` in `CustomerPainter`

**Breaking change**

The default `Semantics` constructor becomes non-const (due to https://github.com/dart-lang/sdk/issues/20962). However, a new `const Semantics.fromProperties` is added that still allowed creating constant `Semantics` widgets ([mailing list announcement](https://groups.google.com/forum/#!topic/flutter-dev/KQXBl2_1sws)).

Fixes https://github.com/flutter/flutter/issues/11791
Fixes https://github.com/flutter/flutter/issues/1666

yjbanov committed:
extract SemanticsProperties out of Semantics

yjbanov committed:
syncing and basic tests

yjbanov committed:
semanticsBuilder getter; shouldRebuildSemantics

yjbanov committed:
implement and test shouldRebuildSemantics logic

goderbauer commented on packages/flutter/lib/src/rendering/custom_paint.dart:
```diff
@@ -0,0 +1,830 @@
+    super.assembleSemanticsNode(node, config, finalChildren);
+  }
+
+  /// Updates `semanticsChildren` from `newSemantics`.
```
What are `semanticsChildren` and `newSemantics`?

yjbanov replied:
Forgot to rename these. Fixed.

goderbauer reviewed (commented):
I like it!

Hixie commented on packages/flutter/lib/src/rendering/custom_paint.dart:
```diff
@@ -0,0 +1,830 @@
+  /// If `oldChild` can be updated, it is updated using [_updateSemanticsChild].
+  /// Otherwise, the node is replaced by a new instance of [SemanticsNode].
+  static bool _canUpdateSemanticsChild(SemanticsNode oldChild, CustomPainterSemantics newSemantics) {
+    return oldChild.key == newSemantics.key;
```
might be best to just inline this

yjbanov replied:
I'd like to keep this because it's used in 7 places and the name `_canUpdateSemanticsChild` makes the purpose of key comparison clearer than `if (oldChild.key == newChild.key)`.

Hixie commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -4696,35 +4696,71 @@ class MetaData extends SingleChildRenderObjectWidget {
+    ),
+  );
+
+  const Semantics.fromProperties({
```
docs

yjbanov replied:
Oops! Done.

Hixie commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -4696,35 +4696,71 @@ class MetaData extends SingleChildRenderObjectWidget {
   ///
   /// The [container] argument must not be null.
-  const Semantics({
+  Semantics({
```
Making this non-const is a breaking change.
We should leave this const.

yjbanov replied:
https://github.com/dart-lang/sdk/issues/20962 prevents me from keeping it `const`. An alternative is to duplicate all the properties, but risking `Semantics` and `SemanticsProperties` going out-of-sync. I also checked with all the code that I could find. There are no `const` instantiations of `Semantics` outside our own unit tests. I actually do not expect `const` instantiation to be very useful, because you at least need a label, which is unlikely to be `const`, or an action, which is even less unlikely to be `const`. So I thought removing `const` but still providing a `const Semantics.fromProperties` would be the right trade-off. Happy to change it to something else though.

Hixie replied:
We should at a minimum announce it on flutter-dev then and get people's feedback, as per our breaking change policy.

yjbanov replied:
True. Sending an announcement shortly.

Hixie commented on packages/flutter/lib/src/foundation/key.dart:
```diff
@@ -0,0 +1,86 @@
+
+  /// Default constructor, used by subclasses.
+  ///
+  /// Useful so that subclasses can call us, because the Key() factory
```
`[new Key]`

yjbanov replied:
Done.

yjbanov commented on packages/flutter/lib/src/semantics/semantics.dart:
```diff
@@ -257,6 +485,9 @@ class SemanticsNode extends AbstractNode with DiagnosticableTreeMixin {
   }
 
+  /// Uniquely identifies this node in the list of sibling nodes.
+  final Key key;
```
Done.

yjbanov commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -4696,35 +4696,71 @@ class MetaData extends SingleChildRenderObjectWidget {
+       assert(properties != null),
        super(key: key, child: child);
 
+  final SemanticsProperties properties;
```
Done.

yjbanov committed:
address comments

yjbanov commented:
Please hold off re-reviewing. Apparently there are more comments that I originally thought.

yjbanov commented on packages/flutter/lib/src/rendering/custom_paint.dart:
```diff
@@ -0,0 +1,830 @@
+  ///
+  /// If the returned function is null, this painter will not contribute new
+  /// [SemanticsNode]s to the semantics tree and the [CustomPaint] corresponding
+  /// to this painter will not create a semantics boundary.
```
Done.

yjbanov commented on packages/flutter/lib/src/rendering/custom_paint.dart:
```diff
@@ -0,0 +1,830 @@
+/// This information is used, for example, by assistive technologies to improve
+/// the accessibility of applications.
+///
+/// Implement [SemanticsPainter.semanticsBuilder] to build the semantic
```
Good catch! Done.

yjbanov commented on packages/flutter/lib/src/rendering/custom_paint.dart:
```diff
@@ -0,0 +1,830 @@
+/// See also:
+///
+/// * [SemanticsNode], which is created using the properties of this class.
+/// * [SemanticsPainter], which creates instances of this class.
```
Done.

yjbanov commented on packages/flutter/test/widgets/custom_painter_test.dart:
```diff
@@ -0,0 +1,631 @@
+      _PainterWithSemantics.semanticsBuilderCallCount = 0;
+    });
+
+    _defineTests();
```
So the tests are not indented too far.

yjbanov committed:
more doc fixes

yjbanov commented:
Addressed all comments (I think!), and sent out [breaking change announcement](https://groups.google.com/forum/#!topic/flutter-dev/KQXBl2_1sws). PTAL.

yjbanov committed:
merge upstream changes

Hixie commented on packages/flutter/lib/src/rendering/custom_paint.dart:
```diff
@@ -0,0 +1,834 @@
+/// To implement a custom painter, either subclass or implement this interface
+/// to define your custom paint delegate. [CustomPaint] subclasses must
+/// implement the [paint] and [shouldRepaint] methods, and may optionally also
+/// implement the [hitTest] method.
```
these docs should talk about the stuff you added

Hixie commented:
One trick for future patches like this: if you move the files in a separate PR than in the PR where you add the feature, the new features becomes easier to review. :-)

![LGTM, but travis is angry.](https://cloud.githubusercontent.com/assets/551196/16323131/36570fa8-395d-11e6-9822-7a0aac54e3b5.png)

yjbanov committed:
adjust the number of time describeSemanticsConfiguration is called

goderbauer reviewed (approved):
LGTM

yjbanov committed:
docs and reexports
